packages/network-controller/src/NetworkController.test.ts

```diff
@@ -1,278 +1,4132 @@
+//       │                 ⋮                                                                               │
+//       │   this.provider.on('error', ...)                                                                │
+//       │                 │                                                                               │
+//       │            verifyNetwork                                                                        │
+//       │                 └─────────────────────────────┐                                                 │
+//       └───────────────────────────────────────────────┼─────────────────────────────────────────────────┘
+//                                                 lookupNetwork
 
 describe('NetworkController', () => {
-  let messenger: NetworkControllerMessenger;
-
   beforeEach(() => {
-    messenger = new ControllerMessenger().getRestricted({
-      name: 'NetworkController',
-      allowedEvents: ['NetworkController:providerConfigChange'],
-      allowedActions: [],
-    });
+    jest.resetAllMocks();
+    jest.useFakeTimers();
   });
 
   afterEach(() => {
-    sinon.restore();
+    jest.useRealTimers();
   });
 
-  it('should set default state', () => {
-    const controller = new NetworkController({
-      messenger,
-      infuraProjectId: 'potate',
+  describe('constructor', () => {
+    it('initializes the state with some defaults', async () => {
+      await withController(({ controller }) => {
+        expect(controller.state).toStrictEqual({
+          network: 'loading',
+          isCustomNetwork: false,
+          providerConfig: { type: 'mainnet' as const, chainId: '1' },
+          properties: { isEIP1559Compatible: false },
+        });
+      });
     });
 
-    expect(controller.state).toStrictEqual({
-      network: 'loading',
-      isCustomNetwork: false,
-      properties: { isEIP1559Compatible: false },
-      providerConfig: {
-        type: 'mainnet',
-        chainId: '1',
-      },
+    it('merges the given state into the default state', async () => {
+      await withController(
+        {
+          state: {
+            isCustomNetwork: true,
+            properties: { isEIP1559Compatible: true },
+          },
+        },
+        ({ controller }) => {
+          expect(controller.state).toStrictEqual({
+            network: 'loading',
+            isCustomNetwork: true,
+            providerConfig: { type: 'mainnet', chainId: '1' },
+            properties: { isEIP1559Compatible: true },
+          });
+        },
+      );
     });
   });
 
-  it('should create a provider instance for default infura network', () => {
-    const networkControllerOpts = {
-      infuraProjectId: 'foo',
-      messenger,
-    };
-    const controller = new NetworkController(networkControllerOpts);
-    controller.providerConfig = {} as ProviderConfig;
-    expect(controller.provider instanceof Web3ProviderEngine).toBe(true);
-  });
+  describe('providerConfig property', () => {
```

Comment:
Sorry if I mistakenly gave this impression (I know I used the word "comprehensively" in the issue), but I'm not sure this was a great use of time in retrospect. I mean, maybe it'll be useful, and having tests for this stuff couldn't hurt I guess. We've learned some things along the way.

But what I had in mind was having comprehensive tests for the behavior that we care about, to ensure it doesn't break. That's in here, but it seems that a lot of these tests are exhaustively testing what this does, rather than what we want it to do.

Reply:
I suppose it wasn't absolutely necessary to be exhaustive, but I think that it was a fruitful exercise in this case. Obviously we are about to make a bunch of improvements Very Soon™️, and when that happens, these tests will go away. But before we really dive in to that, I feel that it is important to establish a baseline that's based on what we have today, not based on what we want tomorrow, or the next day, or the next day. I think this is especially important for _this_ file because it is such a critical controller and because this will be the final version of the controller. That said, I will try in the future to be more aware of how exhaustive we can afford to be when it comes to testing and adjust that balance accordingly.